src/server.rs

```diff
@@ -0,0 +1,294 @@
+        let scuttlebutt_server = scuttlebutt.clone();
+
+        let join_handle = tokio::spawn(async move {
+            let mut server = UdpServer::new(rx, scuttlebutt_server).await?;
+            server.run().await
+        });
+
+        let scuttle_server = Self {
+            channel: tx,
+            scuttlebutt,
+            join_handle,
+        };
+
+        // Advertise the new node to all seed nodes.
+        for node_id in seed_nodes {
+            let _ = scuttle_server.gossip(node_id);
+        }
+
+        scuttle_server
+    }
+
+    /// Call a function with access to the [`ScuttleButt`].
+    pub async fn with_scuttlebutt<F, T>(&self, mut fun: F) -> T
+    where
+        F: FnMut(&ScuttleButt) -> T,
+    {
+        let scuttlebutt = self.scuttlebutt.lock().await;
+        fun(&scuttlebutt)
+    }
+
+    /// Call a function with mutable access to the [`ScuttleButt`].
+    ///
+    /// This will automatically gossip about the changes to all known nodes.
+    pub async fn with_scuttlebutt_mut<F, T>(&self, mut fun: F) -> T
+    where
+        F: FnMut(&mut ScuttleButt) -> T,
+    {
+        let mut scuttlebutt = self.scuttlebutt.lock().await;
+        let retval = fun(&mut scuttlebutt);
+
+        // Broadcast changes to all nodes after mutation.
+        for node in scuttlebutt.cluster_state_map.nodes() {
+            let _ = self.gossip(node);
+        }
+
+        retval
+    }
+
+    /// Shut the server down.
+    pub async fn shutdown(self) -> Result<(), anyhow::Error> {
+        let _ = self.channel.send(ChannelMessage::Shutdown);
+        self.join_handle.await?
+    }
+
+    /// Perform a ScuttleButt "handshake" with another UDP server.
+    pub fn gossip(&self, addr: impl Into<String>) -> Result<(), anyhow::Error> {
+        Ok(self.channel.send(ChannelMessage::Gossip(addr.into()))?)
+    }
+}
+
+/// UDP server for ScuttleButt communication.
+struct UdpServer {
+    channel: UnboundedReceiver<ChannelMessage>,
+    scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    socket: Arc<UdpSocket>,
+}
+
+impl UdpServer {
+    async fn new(
+        channel: UnboundedReceiver<ChannelMessage>,
+        scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    ) -> anyhow::Result<Self> {
+        let socket = {
+            let address = &scuttlebutt.lock().await.self_node_id;
+            Arc::new(UdpSocket::bind(address).await?)
+        };
+
+        Ok(Self {
+            scuttlebutt,
+            channel,
+            socket,
+        })
+    }
+
+    /// Listen for new ScuttleButt messages.
+    async fn run(&mut self) -> anyhow::Result<()> {
+        let mut buf = [0; BUF_SIZE];
+        loop {
+            tokio::select! {
+                result = self.socket.recv_from(&mut buf) => match result {
+                    Ok((len, addr)) => {
+                        let _ = self.process_package(addr, &buf[..len]).await;
+                    }
+                    Err(err) => return Err(err.into()),
```

Comment:
A payload of excessive size is ignored, I've chosen to explicitly add a test for that. But this is still necessary in case something went wrong with the socket. I don't think there's much that can be done as a recovery from this specific error.